subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * TODO:
+ * Support PULL transfer method (from server)
+ */
+
+#define SYS_LOG_DOMAIN "lwm2m_obj_firmware_pull"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_LWM2M_LEVEL
+#include <logging/sys_log.h>
+
+#include <stdio.h>
+#include <net/zoap.h>
+#include <net/net_core.h>
+#include <net/net_pkt.h>
+
+#include "lwm2m_object.h"
+#include "lwm2m_engine.h"
+
+#define STATE_IDLE		0
+#define STATE_CONNECTING	1
+
+#define PACKAGE_URI_LEN				255
+
+static u8_t transfer_state;
+static struct k_work firmware_work;
+static char firmware_uri[PACKAGE_URI_LEN];
+static struct sockaddr firmware_addr;
+static struct net_context *firmware_net_ctx;
+static struct k_delayed_work retransmit_work;
+
+#define NUM_PENDINGS	CONFIG_LWM2M_ENGINE_MAX_PENDING
+#define NUM_REPLIES	CONFIG_LWM2M_ENGINE_MAX_REPLIES
+static struct zoap_pending pendings[NUM_PENDINGS];
+static struct zoap_reply replies[NUM_REPLIES];
+static struct zoap_block_context firmware_block_ctx;
+
+static void
+firmware_udp_receive(struct net_context *ctx, struct net_pkt *pkt, int status,
+		     void *user_data)
+{
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	struct zoap_packet response;
+	int header_len, r;
+	const u8_t *token;
+	u8_t tkl;
+	struct sockaddr from_addr;
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
```

Comment:
I saw similar like of calls in some other function somewhere above, perhaps some refactoring could be done here and common functionality could be separated.